ArduCopter/heli.cpp

```diff
@@ -208,6 +208,27 @@ void Copter::heli_update_autorotation()
             heli_flags.in_autorotation = true;
             motors->set_in_autorotation(heli_flags.in_autorotation);
             motors->set_enable_bailout(true);
+            int32_t gnd_dist = flightmode->get_alt_above_ground_cm();
+
+            // use rangefinder if available
+            switch (rangefinder.status_orient(ROTATION_PITCH_270)) {
+            case RangeFinder::Status::NotConnected:
+            case RangeFinder::Status::NoData:
+            case RangeFinder::Status::OutOfRangeHigh:
+                // use altitude above home for non-functioning rangefinder
+                break;
+
+            case RangeFinder::Status::OutOfRangeLow:
+                // altitude is close to zero (gear should deploy)
+                gnd_dist = 0;
+                break;
+
+            case RangeFinder::Status::Good:
+                // use last good reading
+                gnd_dist = rangefinder_state.alt_cm;
```

Comment:
@Gone4Dirt we had this discussion in the Devcall over a year ago and that is why we developed a complementary filter in order to get away from increasing the rangefinder update rate.  Ferruccio has done substantial testing to ensure this is accurate. 

@Ferruccio1984 i would recommend that you don’t use the suggested variable until you log it and show that it is producing as accurate of data as the complementary filter that we developed.  I think we looked at a similar variable that Tridge had recommended and found it have lag in it.

Reply:
@bnsgeyer Are you referring to the comparison that was posted above between HAGL and the high rate rangefinder measurement?  If so, the method I am suggesting is not the same as XKF5.HAGL.
@Ferruccio1984 if you have already looked at comparing the output from get_rangefinder_height_interpolated_cm() and the complimentary filter you implemented could you share a plot of that data comparison please?  If this has not been compared then it would be great to log and compare this as Bill suggests.

Reply:
@Gone4Dirt yes, that is what I am referring to. No one even mentioned this rangefinder interpolation parameter back then.  Is it new? Do you know it’s implemented?  Are they also using a complementary filter?

Reply:
> If this has not been compared then it would be great to log and compare this as Bill suggests.

@Gone4Dirt I haven't compared "get_rangefinder_height_interpolated_cm()" with the resutls from the complimentary filter in the tests I have done so far. I have now incorporated this (for logging only) in the current AC 4.4.4 I'm using, so we can evaluate.

Reply:
@bnsgeyer @Gone4Dirt flight tested today, with logged altitude estimate from your suggestion. Graph below compares  ground distance estimate from autorotation filter("Hest"), interpolated altitude (obtained from get_rangefinder_height_interpolated_cm() method "Ialt") and pure rangefinder measurement:
![image](https://github.com/ArduPilot/ardupilot/assets/8063851/723a768f-e34a-4a34-bea4-c4fa4e26574d)
onboard video here: https://www.youtube.com/watch?v=Xw7XLro6raU

Reply:
@Ferruccio1984, that's a really cool video.

Reply:
@Gone4Dirt @Ferruccio1984 I did a little digging on the variable get_rangefinder_height_interpolated_cm().  It appears that it is calculated using the filtered value of rangefinder (rangefinder_state.alt_cm_filt).   The default value of this is 0.5 hz which explains the lagged value in ferruccio's test.  I think if you set the parameter for the filter frequency to zero then you may get a better correlation.   

The complementary filter in the autorotation code uses descent rate instead of inertial altitude to interpolate.  It is possible that there will be discontinuities in altitude with this method if the LPF is disabled (param set to zero).  I suppose it could be smoothed by using a higher LPF cutoff frequency.   

@Gone4Dirt I have concerns with using this variable. It is possible that it will be fine but if other Dev's change how this is calculated without our knowledge then it could cause problems.  Also now the user will have to potentially tune this by changing the LPF cutoff frequency.  What is the benefit of using this variable?

Reply:
@Ferruccio1984 thanks for doing that testing and comparison.  Much appreciated.

@bnsgeyer my main concern was to avoid code duplication.  If there is a notable difference in performance that justifies the need for the new method then it should go in. However, having two different functions that result in the same outcome is wasteful and can be confusing to others that come to the code in a few years time, hence why I wanted to bottom this out.

Regarding concerns that the function will be broken without our knowledge: a credible concern but the fix to that is to write autotests (of which I am happy to help write).  Putting the function in autorotation library does not assure it will not get broken (quite the reverse in my experience) as it is still dependant on downstream inputs from rangefinder etc.

@Ferruccio1984 is the plan to use this complimentary filter with other height measuring devices in the future (E.g. Radar) or is this likely to remain lidar specific?  The reason I ask is because as the code stands at the moment is only takes lidar and IMHO we should move this filter to rangefinder.

Reply:
> @Ferruccio1984, that's a really cool video.

thank you @rmackay9 !

Reply:
> is the plan to use this complimentary filter with other height measuring devices in the future (E.g. Radar) or is this likely to remain lidar specific?

@Gone4Dirt good question. I think it would be desirable to have this filter just downstream of whatever sensor we might want to use.  

> thanks for doing that testing and comparison. Much appreciated.

My Pleasure!

Reply:
@Gone4Dirt Ferruccio used a rangefinder filter cutoff frequency of 0 to disable the filter and the height estimates are nearly the same.  Ferruccio brought up a good question when we were discussing the results.  How would we use this existing variable in practice.  The user maybe using the rangefinder for some other purpose where the filter cutoff frequency would be a lower value or the default.  Would the autorotation code change that cutoff frequency if the aircraft experiences an engine failure.  or would we say that this would be a dedicated rangefinder for autorotations? So I just want to throw it out there for discussion.  We don't want to duplicate code but the employment of the rangefinder and this variable requires different settings for autorotations.  I suppose we could make our own instance using the rangefinder methods? Then we'd be able to set the cutoff frequency separately.

Reply:
@bnsgeyer Ok, this is pretty cool.  It is a fair comment but I can't think of a reason why a user would set a lower filter cutoff freq.  By that I mean, if the provided data is good enough for use in autorotation there is no reason to filter it any more.  The reason the filter (as far as I am aware) is there is because LiDaR measurements tend to be quite erratic.  But this is a problem with the sensor in general not an application specific requirement.

Not sure what is the best approach to setting the filter.  I am not a big fan of setting the filter frequency for the users automatically (E.g. On init of autorotation we could set the param).
I am also not a fan of another rangefinder instance as I don't think it is necessary.  My preference is simply clear instructions on the autorotation setup wiki to set this filter and an explanation of why it is important.